### PR TITLE
feat(pr-baseline-2): Support workflow enforcement in GitHub repos

### PR DESCRIPTION
**Pull request by DaveTryon**
#### Details

The workflow enforcement code was in a class that only ran if the code was hosted in ADO, not in GitHub. This PR splits workflow-related part of the `AdoPullRequestCommentCreator` class into a new `WorkflowEnforcer` class. The IOC container is configured to create the classes so that a `WorkflowEnforcer` object is always included in the collection of `ProgressReporter` objects. If the code happens to be hosted in ADO, then we also include an `AdoPullRequestCommentCreator` before the `WorkflowEnforcer`, since we want to update the PR comment.

Since `WorkflowEnforcer` will _always_ be the last item in the chain, it's the _only_ class that should do anything that would cause the task to intentionally fail. `WorkflowEnforcer` also runs in all scenarios, regardless of the value for [Build.Reason](https://docs.microsoft.com/en-us/javascript/api/azure-devops-extension-api/buildreason)

##### Motivation

PR baselining feature

##### Context

This is a PR where I wouldn't recommend reviewing the individual commits. I took a bit of a wrong turn regarding the design then realized that it shouldn't be there.

In the interest of unblocking validation, this PR does not contain update the errors to display using the format described at https://docs.microsoft.com/en-us/azure/devops/pipelines/scripts/logging-commands?view=azure-devops&tabs=bash. We can consider that as a separate PR.

#### Pull request checklist

- [x] Addresses an existing issue: #920
- [x] Added relevant unit test for your changes. (`yarn test`)
- [x] Verified code coverage for the changes made. Check coverage report at: `<rootDir>/test-results/unit/coverage`
- [x] Ran precheckin (`yarn precheckin`)
- [x] (after PR created) The `Accessibility Checks (pull_request)` check should fail. All other checks should pass.

### CODE REVIEW

File: packages/ado-extension/src/progress-reporter/enforcement/workflow-enforcement.ts [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+    }
+
+    private failOnBaselineNotUpdated(suggestedBaselineUpdate: null | BaselineFileContent): void {
+        if (this.adoTaskConfig.getBaselineFile() && suggestedBaselineUpdate) {
+            throw new Error('Failed: The baseline file needs to be updated. See the PR comments for more details.');
+        }
+    }
+
+    private isSupported(): boolean {
+        return this.adoTask.getVariable('Build.Reason') == 'PullRequest';
```
Comment (lisli1) [inline packages/ado-extension/src/progress-reporter/enforcement/workflow-enforcement.ts:64]: If we have this for the isSupported condition, does that mean it won't enforce workflows in CI builds?

**Review by lisli1 — APPROVED** [review]

**Review by jalkire — APPROVED** [review]
👍

File: packages/ado-extension/src/progress-reporter/enforcement/workflow-enforcement.ts (hunk 1 above)
Comment (DaveTryon) [reply packages/ado-extension/src/progress-reporter/enforcement/workflow-enforcement.ts:64]: As discussed, we'll just remove the isSupported so the enforcer always runs

File: packages/ado-extension/src/progress-reporter/enforcement/workflow-enforcer.ts [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+import { inject, injectable } from 'inversify';
+import { ProgressReporter } from '@accessibility-insights-action/shared';
+import { CombinedReportParameters } from 'accessibility-insights-report';
+import * as AdoTask from 'azure-pipelines-task-lib/task';
+import { BaselineEvaluation, BaselineFileContent } from 'accessibility-insights-scan';
+
+@injectable()
+export class WorkflowEnforcer extends ProgressReporter {
+    public static readonly CURRENT_COMMENT_TITLE = 'Results from Current Run';
+    public static readonly PREVIOUS_COMMENT_TITLE = 'Results from Previous Run';
```
Comment (lisli1) [inline packages/ado-extension/src/progress-reporter/enforcement/workflow-enforcer.ts:15]: Just noticed this -- is this left over from copy/paste?

Comment (DaveTryon) [reply packages/ado-extension/src/progress-reporter/enforcement/workflow-enforcer.ts:15]: Yes. Fixing

File: packages/ado-extension/src/progress-reporter/enforcement/workflow-enforcer.ts [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+    }
+
+    public async start(): Promise<void> {
+        // We don't do anything for workflow enforcement
+    }
+
+    public async completeRun(combinedReportResult: CombinedReportParameters, baselineEvaluation?: BaselineEvaluation): Promise<void> {
+        this.failIfAccessibilityErrorExists(combinedReportResult);
+        this.failIfBaselineNeedsUpdating(baselineEvaluation);
+        return Promise.resolve();
```
Comment (dbjorge) [inline packages/ado-extension/src/progress-reporter/enforcement/workflow-enforcer.ts:28]: nit: This line should be unnecessary (the `async` part of the message signature implies the same behavior if you just let the function end)

File: packages/ado-extension/src/progress-reporter/enforcement/workflow-enforcer.ts [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+
+    private failIfAccessibilityErrorExists(combinedReportResult: CombinedReportParameters): void {
+        if (this.adoTaskConfig.getFailOnAccessibilityError() && combinedReportResult.results.urlResults.failedUrls > 0) {
+            throw new Error('Failed Accessibility Error');
+        }
+    }
+
+    private failIfBaselineNeedsUpdating(baselineEvaluation?: BaselineEvaluation): void {
+        if (baselineEvaluation && this.adoTaskConfig.getBaselineFile() && baselineEvaluation.suggestedBaselineUpdate) {
+            throw new Error('Failed: The baseline file needs to be updated. See the PR comments for more details.');
```
Comment (dbjorge) [inline packages/ado-extension/src/progress-reporter/enforcement/workflow-enforcer.ts:44]: This should probably be generalized a bit, since it can occur in a non-PR environment

File: packages/ado-extension/src/progress-reporter/enforcement/workflow-enforcer.ts [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+import { ProgressReporter } from '@accessibility-insights-action/shared';
+import { CombinedReportParameters } from 'accessibility-insights-report';
+import * as AdoTask from 'azure-pipelines-task-lib/task';
+import { BaselineEvaluation, BaselineFileContent } from 'accessibility-insights-scan';
+
+@injectable()
+export class WorkflowEnforcer extends ProgressReporter {
+    constructor(
+        @inject(ADOTaskConfig) private readonly adoTaskConfig: ADOTaskConfig,
+        @inject(AdoIocTypes.AdoTask) private readonly adoTask: typeof AdoTask,
```
Comment (dbjorge) [inline packages/ado-extension/src/progress-reporter/enforcement/workflow-enforcer.ts:16]: This looks unused

File: packages/ado-extension/src/progress-reporter/enforcement/workflow-enforcer.ts (hunk 3 above)
Comment (DaveTryon) [reply packages/ado-extension/src/progress-reporter/enforcement/workflow-enforcer.ts:28]: It was add the `Promise.resolve` or add a lint suppression. I chose the former, but I can change it

File: packages/ado-extension/src/ioc/setup-ioc-container.ts [hunk 6]
```diff
@@ -10,26 +10,25 @@ import { AdoPullRequestCommentCreator } from '../progress-reporter/pull-request/
...
         .bind(iocTypes.ProgressReporters)
         .toDynamicValue((context) => {
             switch (process.env.BUILD_REPOSITORY_PROVIDER) {
                 case 'TfsGit': {
                     const pullRequestCommentCreator = context.container.get(AdoPullRequestCommentCreator);
-                    return [pullRequestCommentCreator];
-                }
-                case 'GitHub': {
-                    return [];
+                    return [pullRequestCommentCreator, context.container.get(WorkflowEnforcer)];
```
Comment (dbjorge) [inline packages/ado-extension/src/ioc/setup-ioc-container.ts:28]: It feels a bit sketchy that an exception in the PR comment creator would cause the workflow enforcer not to run. I'd be okay with fixing that in a separate PR, but for now it would be nice to at least leave a comment here explaining why the order of this list is important

File: packages/ado-extension/src/progress-reporter/enforcement/workflow-enforcer.spec.ts [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+        it('succeeds in happy path (baseline not enabled)', async () => {
+            const reportStub = {} as CombinedReportParameters;
+            const baselineEvaluationStub = {} as BaselineEvaluation;
+
+            setupFailOnAccessibilityError(false);
+            setupBaselineFileParameterDoesNotExist();
+
+            workflowEnforcer = buildWorkflowEnforcerWithMocks();
+
+            await workflowEnforcer.completeRun(reportStub, baselineEvaluationStub);
```
Comment (dbjorge) [inline packages/ado-extension/src/progress-reporter/enforcement/workflow-enforcer.spec.ts:95]: I think in the "baseline not enabled" case, the expected input would be for no evaluation to be passed, not an empty evaluation

File: packages/ado-extension/src/progress-reporter/enforcement/workflow-enforcer.ts (hunk 5 above)
Comment (DaveTryon) [reply packages/ado-extension/src/progress-reporter/enforcement/workflow-enforcer.ts:16]: Oops! Remnant of when we had IsSupported in this class

**Review by dbjorge — APPROVED** [review]
